assignment4/count_islands.py

```diff
@@ -0,0 +1,128 @@
+    It then counts the roots of trees that represent land tiles to 
+    calculate the islands.
+    
+    NB. True -> Land tile
+        False -> Water tile
+    NB2. Assume rowMax and colMax are not out of bounds of the array
+        
+        Args:
+            rowMax: an integer, the number of rows in the array
+            colMax: an integer, the number of columns in the array
+            tiles: a two-dimensional array of boolean values
+        
+        Returns:
+            The number of islands found in the array.
+    """
+    if tiles is None or len(tiles) == 0 or len(tiles[0]) == 0: 
+        raise ValueError("Please insert a non-empty array of tiles.")
+    
+    dis_set = DisjointSet(rowMax * colMax)
+    
+    for i in range(rowMax):
+        for j in range(colMax):
+            if tiles[i][j] == False: continue
+            
+            # right
+            if in_bound(j+1, colMax) and tiles[i][j+1] == True:
+                dis_set.merge(i*rowMax+j, i*rowMax+j+1)
+            # left
+            if in_bound(j-1, colMax) and tiles[i][j-1] == True:
+                dis_set.merge(i*rowMax+j, i*rowMax+j-1)
+            # top
+            if in_bound(i+1, rowMax) and tiles[i+1][j]:
+                dis_set.merge(i*rowMax+j, (i+1)*rowMax+j)
+            # bottom
+            if in_bound(i-1, rowMax) and tiles[i-1][j]:
+                dis_set.merge(i*rowMax+j, (i-1)*rowMax+j)
+    
+    visited_roots = []
+    island_counter = 0
+    for i in range(rowMax):
+        for j in range(colMax):
+            if tiles[i][j] == True and dis_set.find(i*rowMax+j) not in visited_roots:
+                    root = dis_set.find(i*rowMax+j)
+                    island_counter = island_counter + 1
+                    visited_roots.append(root)
+    
+    return island_counter            
+    
+def count_islands_rec(rowMax, colMax, tiles):
+    """Counts the number of islands that exist in the given array.
+    
+    This approach uses DFS and Flood-Fill.
+    
+    NB. True -> Land tile
+        False -> Water tile
+    NB2. Assume rowMax and colMax are not out of bounds of the array
+    
+        Args:
+            rowMax: an integer, the number of rows in the array
+            colMax: an integer, the number of columns in the array
+            tiles: a two-dimensional array of boolean values
+        
+        Returns:
+            The number of islands found in the array.
+    """
+    if tiles is None or len(tiles) == 0 or len(tiles[0]) == 0: 
+        raise ValueError("Please insert a non-empty array of tiles.")
+    
+    counter = 0
+    visited = [None] * rowMax
+    for i in range(rowMax):
+        visited[i] = [False] * colMax
+    
+    for i in range(rowMax):
+        for j in range(colMax):
+            if visited[i][j]:
+                continue
+            elif not tiles[i][j]:
+                visited[i][j] = True
+                continue
+            counter += 1
```

Comment:
I find the use of `continue` in this method a bit confusing.

I think continue works well when you want to interrupt the flow, especially when a whole block does not apply. In this case, the first one works reasonably well in this case. However, you also have an `elif`, which is unnecessary if you have a `continue` on the previous branch. And then the last case does not have the `else`.

There is no fast rule that applies to this, but I would suggest one of the following:
```python
  if visited[i][j]:
    continue

  if not tiles[i][j]:  # This is a water tile, so no island there.
    visited[i][j] = True
    continue

  counter += 1
  floor_fill(rowMax, volMax, i, j, tiles, visited)
```
Notice that I used continue but never used `else` in this case.

Alternatively:
```python
  if visited[i][j]:
    continue

  if not tiles[i][j]:
    visited[i][j] = True
  else:
    counter += 1
    flood_fill(rowMax, colMax, i, j, tiles, visited)
```
In this case, I used the `else` for the last two (and not the `continue` there).

I think it is about how you want the reader to think about the code.

In the first example, there are two `exceptions` and one normal case. In the second example, there is an exception (the tile is already visited) and two normal cases, with different actions.

I personally prefer the latter one, because that way the `continue` is only used for those cases where there is nothing at all to be done.